internal/core/src/clustering/MemClustering.cpp

```diff
@@ -0,0 +1,360 @@
+        return num_vectors;
+    };
+    std::vector<milvus::proto::segcore::ClusteringCentroidIdMappingStats>
+        stats_arr;
+    int64_t cur_offset = 0;
+    for (auto segment_id : segment_ids) {
+        milvus::proto::segcore::ClusteringCentroidIdMappingStats stats;
+        auto num_offset = num_row_map.at(segment_id);
+        for (auto j = 0; j < num_offset; j++) {
+            stats.add_centroid_id_mapping(centroid_id_mapping[cur_offset + j]);
+        }
+        auto num_vectors = compute_num_in_centroid(
+            centroid_id_mapping, cur_offset, cur_offset + num_offset);
+        for (uint64_t j = 0; j < num_clusters_; j++) {
+            stats.add_num_in_centroid(num_vectors[j]);
+        }
+        cur_offset += num_offset;
+        stats_arr.emplace_back(stats);
+    }
+    return stats_arr;
+}
+
+template <typename T>
+void
+MemClustering<T>::StreamingAssignandUpload(
+    const T* centroids,
+    const std::map<int64_t, std::vector<std::string>>& insert_files,
+    const std::map<int64_t, int64_t>& num_rows) {
+    LOG_INFO("start upload");
+    auto byte_size = centroid_stats_.ByteSizeLong();
+    std::unique_ptr<uint8_t[]> data = std::make_unique<uint8_t[]>(byte_size);
+    centroid_stats_.SerializeToArray(data.get(), byte_size);
+    std::unordered_map<std::string, int64_t> remote_paths_to_size;
+    LOG_INFO("start upload cluster centroids file");
+    AddClusteringResultFiles(
+        file_manager_->GetChunkManager().get(),
+        data.get(),
+        byte_size,
+        GetRemoteCentroidsObjectPrefix() + "/" + std::string(CENTROIDS_NAME),
+        remote_paths_to_size);
+    cluster_result_.centroid_path =
+        GetRemoteCentroidsObjectPrefix() + "/" + std::string(CENTROIDS_NAME);
+    cluster_result_.centroid_file_size =
+        remote_paths_to_size.at(cluster_result_.centroid_path);
+    remote_paths_to_size.clear();
+    LOG_INFO("upload cluster centroids file done");
+
+    auto serializeIdMappingAndUpload =
+        [&](const int64_t segment_id,
+            milvus::proto::segcore::ClusteringCentroidIdMappingStats&
+                id_mapping_pb) {
+            auto byte_size = id_mapping_pb.ByteSizeLong();
+            std::unique_ptr<uint8_t[]> data =
+                std::make_unique<uint8_t[]>(byte_size);
+            id_mapping_pb.SerializeToArray(data.get(), byte_size);
+            AddClusteringResultFiles(
+                file_manager_->GetChunkManager().get(),
+                data.get(),
+                byte_size,
+                GetRemoteCentroidIdMappingObjectPrefix(segment_id) + "/" +
+                    std::string(OFFSET_MAPPING_NAME),
+                remote_paths_to_size);
+        };
+
+    LOG_INFO("start upload cluster id mapping file");
+    for (auto& [segment_id, segment_files] : insert_files) {
+        // id mapping has been computed, just upload to remote
+        if (trained_segmentid_to_offset_.find(segment_id) !=
+            trained_segmentid_to_offset_.end()) {
```

Comment:
filed is for variables that will be used in different API call. This `trained_segmentid_to_offset_` is only used in `Run()`, so we can make it a local variable.
Plz check other variables.